content/blog/single-message-transforms-swiss-army-knife-of-kafka-connect.asciidoc

```diff
@@ -0,0 +1,235 @@
+
+== Tombstone Handling
+
+Tombstone records are Kafka records with a `null` value.
+They carry special semantics when working with https://kafka.apache.org/documentation/#compaction[compacted topics]:
+during log compaction, all records with the same key as a tombstone record will be removed from the topic.
+
+Tombstones will be retained on a topic for a configurable time before compaction happens (controlled via https://kafka.apache.org/documentation/#delete.retention.ms[`delete.retention.ms`] topic setting),
+which means that also Kafka Connect sink connectors need to handle them.
+Unfortunately though, not all connectors are prepared for records with a `null` value,
+typically resulting in ``NullPointerException``s and similar.
+A filtering SMT such as the one above can be used to drop tombstone records in such case.
+
+But also the exact opposite -- producing tombstone records -- can be useful:
+some sink connectors use tombstone records as the indicator to delete corresponding rows from a downstream datastore.
+Now when using a a CDC connector like Debezium to capture changes from a database where "soft deletes" are used (i.e. records are not physically deleted but a logically deleted flag is set to `true` when deleting a record), those change events will be exported as update events (which they technically are).
+A bespoke SMT can be used to translate these update events into tombstone records, triggering the deletion of corresponding records in downstream datastores.
+
+_Mac's rating:_ &nbsp;&nbsp; 📎📎📎📎&nbsp;&nbsp; _SMTs work well to discard tombstones or convert soft delete events into tombstones. What's not possible though is to keep the original event and produce an additional tombstone record at the same time_
+
+== Externalizing Large Payloads
+
+Even some advanced https://www.enterpriseintegrationpatterns.com/[enterprise application patterns] can be implemented with the help of SMTs, one example being the https://www.enterpriseintegrationpatterns.com/patterns/messaging/StoreInLibrary.html[claim check pattern].
+This pattern comes in handy in situations like this:
+
+[quote]
+____
+A message may contain a set of data items that may be needed later in the message flow, but that are not necessary for all intermediate processing steps. We may not want to carry all this information through each processing step because it may cause performance degradation and makes debugging harder because we carry so much extra data.
+
+-- Gregor Hohpe, Bobby Woolf; Enterprise Application Patterns
+____
+
+A specific example could again be a CDC connector that captures changes from a database table `Users`, with a BLOB column that contains the user's profile picture
+(surely not a best practice, still not that uncommon in reality...).
+
+When propagating change data events from that table to Apache Kafka,
+adding the picture data to each event poses a significant overhead.
```

Comment:
I would even add here that you can face payloads containing blobs that are simply too big. remember that payloads which clearly exceed 10MB are absolutely not recommended. there are messaging / streaming solutions which have even strict payload size limits of < 1 MB. in the latter case the claim check pattern becomes vital to be able to handle such payloads at all.

Reply:
Good idea, added an "info" box on this.